### ISSUE TITLE
Import from desktop crashed with any_cast exception

### ISSUE THREAD

**Issue by sesse** [bepaald/signalbackup-tools#163]
Hi,

I'm having an old-ish backup that I'm trying to backfill with desktop data (obviously censoring the passphrase):

```
> ./signalbackup-tools signal-2023-06-23-11-10-01.backup 'XXXXX XXXXX …' --importfromdesktop --output merged-db.backup
signalbackup-tools (./signalbackup-tools) source version 20231123.165937
BACKUPFILE VERSION: 0
BACKUPFILE SIZE: 14207470986
COUNTER: 336956965
Reading backup file...
FRAME 332573 (100.0%)... Read entire backup file...

done!
Database version: 196
Trying to match conversation (1/195) (type: private)
 - Importing 3 messages into thread._id 25
Trying to match conversation (2/195) (type: private)
 - Importing 7012 messages into thread._id 19
Warning: Failed to get number of attachments in quoted message. Skipping
Warning: Failed to get number of attachments in quoted message. Skipping
Warning: Failed to get number of attachments in quoted message. Skipping
Warning: Failed to get number of attachments in quoted message. Skipping
Warning: Failed to get number of attachments in quoted message. Skipping
Warning: Failed to get number of attachments in quoted message. Skipping
Warning: Failed to get number of attachments in quoted message. Skipping
Warning: Failed to get number of attachments in quoted message. Skipping
Warning: Failed to get number of attachments in quoted message. Skipping
Warning: Failed to get number of attachments in quoted message. Skipping
Warning: Failed to get number of attachments in quoted message. Skipping
Warning: Failed to get number of attachments in quoted message. Skipping
Warning: Failed to get number of attachments in quoted message. Skipping
Trying to match conversation (3/195) (type: private)
 - Importing 7 messages into thread._id 235
Trying to match conversation (4/195) (type: private)
 - Importing 6 messages into thread._id 180
Skipping conversation, conversation has no messages (5/195)
Trying to match conversation (6/195) (type: private)
 - Importing 57 messages into thread._id 96
Trying to match conversation (7/195) (type: private)
 - Importing 639 messages into thread._id 146
Trying to match conversation (8/195) (type: private)
Warning: Failed to find matching thread for conversation, creating. (id: 3c7761f6-34db-4fb6-b42e-67edb77b57fcWarning: Table 'thread' does not contain any column 'active'. Removing
, thread_id: 752)
 - Importing 20 messages into thread._id 752
Trying to match conversation (9/195) (type: private)
 - Importing 8048 messages into thread._id 11
Trying to match conversation (10/195) (type: private)
 - Importing 81 messages into thread._id 278
Trying to match conversation (11/195) (type: private)
Warning: Failed to determine uuid. Trying with phone number...
 - Importing 77 messages into thread._id 43
Trying to match conversation (12/195) (type: private)
 - Importing 11 messages into thread._id 174
Trying to match conversation (13/195) (type: private)
 - Importing 9 messages into thread._id 236
Trying to match conversation (14/195) (type: private)
 - Importing 10271 messages into thread._id 280
Trying to match conversation (15/195) (type: private)
 - Importing 7 messages into thread._id 215
Trying to match conversation (16/195) (type: private)
 - Importing 16 messages into thread._id 245
Trying to match conversation (17/195) (type: private)
 - Importing 1 messages into thread._id 121
Trying to match conversation (18/195) (type: private)
 - Importing 79 messages into thread._id 117
Getting value of field 'sent_at' (idx 10). Value as string: {"high":350,"low":-829903741,"unsigned":true}
Type: NSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE Requested type: x
terminate called after throwing an instance of 'std::bad_any_cast'
  what():  bad any_cast
```

I assume NSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE is std::string. x is a bit weirder, but perhaps long long int?

Importing and reexporting the phone database does not help; I suppose this value is coming from  the desktop database. To me, this looks like something like the “Long” interface from protobuf.js, in JSON form. So perhaps it should be interpreted as (350 << 32) | (0xce88ac83), aka 1506703617155, which divided by 1e6 would be some point at September 29th 2017? That's not impossible at all (my database goes pretty far back).

**Comment by bepaald** [comment]
>I assume NSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE is std::string. x is a bit weirder, but perhaps long long int?

>Importing and reexporting the phone database does not help; I suppose this value is coming from the desktop database. To me, this looks like something like the “Long” interface from protobuf.js, in JSON form. So perhaps it should be interpreted as (350 << 32) | (0xce88ac83), aka 1506703617155, which divided by 1e6 would be some point at September 29th 2017? That's not impossible at all (my database goes pretty far back).

Wow, that's a new one. I think you're right in your theory, but I've never seen numerical values stored like that. With the info you've provided, it shouldn't be too hard to fix. I'm not at home right now, but I'll get on it tomorrow as soon as I am.

I'm just a bit worried: there are of course a lot of places where I query (and expect) a number. I hope this format isn't used all over your database. You'll have to tell me I'm afraid.

Thanks for reporting, I'll get back to you tomorrow!

**Comment by sesse** [comment]
> I'm just a bit worried: there are of course a lot of places where I query (and expect) a number.

I assume that valueAsInt() wouldn't mask this issue and return 0 instead or something?

This patch, while obviously unfit for upstream, worked for me:

```diff
index 310baba..72c9232 100644
--- a/sqlitedb/sqlitedb.h
+++ b/sqlitedb/sqlitedb.h
@@ -661,6 +661,15 @@ inline T SqliteDB::QueryResults::getValueAs(size_t row, std::string const &heade
   {
     std::cout << "Getting value of field '" << header << "' (idx " << i << "). Value as string: " << valueAsString(row, i) << std::endl;
     std::cout << "Type: " << d_values[row][i].type().name() << " Requested type: " << typeid(T).name() << std::endl;
+    if (d_values[row][i].type() == typeid(std::string) && std::is_same_v<T, long long int>) {
+      long long int high, low;
+      if (sscanf(valueAsString(row, i).c_str(), "{\"high\":%lld,\"low\":%lld,\"unsigned\":true}", &high, &low) == 2) {
+        long long int val = (high << 32) | (unsigned int)(low);
+        std::cout << "HACK: value of " << valueAsString(row, i) << " is " << val << std::endl;
+        return std::any_cast<T>(val);
+      }
+    }
+    abort();
     //return T{};
   }
   return std::any_cast<T>(d_values[row][i]);
```

I got about 30 instances in my database, all in the same conversation. (I don't know why I didn't need to divide by 1000. Perhaps somehow my first calculation was wrong.) I also got one of these, which was nonfatal:

```
Warning : Got empty author uuid, here is some additional info:
json_extract(json, '$.reactions')
[{"emoji":"❤️","remove":false,"targetAuthorE164":null,"targetTimestamp":1602867497418,"timestamp":1602867504000,"fromId":"13338cbe-8445-418c-b073-a8a0c24bf732"}]
Error: Asked to find recipient._id for empty uuid. Refusing
```

And these:

```
Trying to match conversation (82/195) (type: private)
 - Importing 27687 messages into thread._id 2
Error: Trying to set attachment data. Skipping.
Pending: 0
Error: Trying to set attachment data. Skipping.
Pending: 0
Error: Trying to set attachment data. Skipping.
Pending: 0
```

And a couple more. It's still running, I'll let you know whether it worked in the end :-)

**Comment by sesse** [comment]
It completed, but it got way too big; basically the sum of size of the mobile + desktop dumps (almost 30 GB). So I assume something went wrong in the deduplication; about 90% of the history should be shared between the two.

**Comment by sesse** [comment]
(Also, Signal crashed when trying to start after restoring the backup; I'll need to try it again and then capture the logs)

**Comment by sesse** [comment]
Here we are; a couple of these during import:

```
12-02 23:08:00.151  8349  8414 W FullBackupImporter: Bad MAC for attachment 1487958796660! Can't restore it.
12-02 23:08:00.151  8349  8414 W FullBackupImporter: org.thoughtcrime.securesms.backup.BackupRecordInputStream$BadMacException
12-02 23:08:00.151  8349  8414 W FullBackupImporter: 	at org.thoughtcrime.securesms.backup.BackupRecordInputStream.readAttachmentTo(BackupRecordInputStream.java:169)
12-02 23:08:00.151  8349  8414 W FullBackupImporter: 	at org.thoughtcrime.securesms.backup.FullBackupImporter.processAttachment(FullBackupImporter.java:203)
12-02 23:08:00.151  8349  8414 W FullBackupImporter: 	at org.thoughtcrime.securesms.backup.FullBackupImporter.importFile(FullBackupImporter.java:120)
12-02 23:08:00.151  8349  8414 W FullBackupImporter: 	at org.thoughtcrime.securesms.backup.FullBackupImporter.importFile(FullBackupImporter.java:91)
12-02 23:08:00.151  8349  8414 W FullBackupImporter: 	at org.thoughtcrime.securesms.registration.fragments.RestoreBackupFragment$2.doInBackground(RestoreBackupFragment.java:296)
12-02 23:08:00.151  8349  8414 W FullBackupImporter: 	at org.thoughtcrime.securesms.registration.fragments.RestoreBackupFragment$2.doInBackground(RestoreBackupFragment.java:281)
12-02 23:08:00.151  8349  8414 W FullBackupImporter: 	at android.os.AsyncTask$3.call(AsyncTask.java:394)
12-02 23:08:00.151  8349  8414 W FullBackupImporter: 	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
12-02 23:08:00.151  8349  8414 W FullBackupImporter: 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:305)
12-02 23:08:00.151  8349  8414 W FullBackupImporter: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
12-02 23:08:00.151  8349  8414 W FullBackupImporter: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
```

And then Signal dies completely with:

```
12-02 23:08:01.512  8349  8414 E SignalUncaughtException: java.lang.RuntimeException: An error occurred while executing doInBackground()^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at android.os.AsyncTask$4.done(AsyncTask.java:415)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:381)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at java.util.concurrent.FutureTask.setException(FutureTask.java:250)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at java.util.concurrent.FutureTask.run(FutureTask.java:269)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:305)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at java.lang.Thread.run(Thread.java:1012)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException: Caused by: java.lang.ArrayIndexOutOfBoundsException: src.length=0 srcPos=-10 dst.length=10 dstPos=0 length=10^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at java.lang.System.arraycopy(System.java:527)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at org.thoughtcrime.securesms.backup.BackupRecordInputStream.readFrame(BackupRecordInputStream.java:184)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at org.thoughtcrime.securesms.backup.BackupRecordInputStream.readFrame(BackupRecordInputStream.java:96)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at org.thoughtcrime.securesms.backup.FullBackupImporter.importFile(FullBackupImporter.java:113)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at org.thoughtcrime.securesms.backup.FullBackupImporter.importFile(FullBackupImporter.java:91)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at org.thoughtcrime.securesms.registration.fragments.RestoreBackupFragment$2.doInBackground(RestoreBackupFragment.java:296)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at org.thoughtcrime.securesms.registration.fragments.RestoreBackupFragment$2.doInBackground(RestoreBackupFragment.java:281)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at android.os.AsyncTask$3.call(AsyncTask.java:394)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       at java.util.concurrent.FutureTask.run(FutureTask.java:264)^M
12-02 23:08:01.512  8349  8414 E SignalUncaughtException:       ... 4 more^M
```

**Comment by sesse** [comment]
OK, I fiddled with --autolimitdates so that it handled the desktop data being newer than the mobile backup (it only handled the other way), and that seems to have done it? No crash on import, and no huge database. There are 3–4 messages and missed calls that are phantom unread that were not before, and all of my conversations are starting on June, but it seems like things are generally back on track.

**Comment by bepaald** [comment]
>I assume that valueAsInt() wouldn't mask this issue and return 0 instead or something?

I'm afraid it will, yes...

>This patch, while obviously unfit for upstream, worked for me:

Indeed not how I would do it, mostly because the json does not have a guaranteed format like that, it could switch the keys or leave some out (for default values). But it seems to have worked for your case, so that's good.

I've created a custom sqlite function to retrieve the jsonlong, which enabled me to make use of sqlite's built-in json parsing abilities. But, it does require me to explicitly `SELECT JSONLONG(sent_at)` instead of just `SELECT sent_at`.

>I got about 30 instances in my database, all in the same conversation. (I don't know why I didn't need to divide by 1000. Perhaps somehow my first calculation was wrong.) 

Timestamps in Signal (Android and desktop) are stored as milliseconds since epoch, not seconds.

>I also got one of these, which was nonfatal:

That's a weird one, it seems the author of the emoji-reaction is not found. Would it be possible for you to open your Signal Desktop instance and find this message? The reaction was posted October 16 2020, 18:58:24 CEST (but the message could be older). If this is too difficult (absolutely possible, if you have many active conversations around that period), I could write some instructions on how to more easily find it by running some queries on the database.

>And these:

Generally in these cases the attachment is missing (may have been deleted), I added a little more output in this case. Your database seems to be a bit different though, maybe the attachments' filename is simply under a different json-key? If you want to investigate further, let me know and I'll have the program output some more info at this point.

>So I assume something went wrong in the deduplication

Ehm... what deduplication? :)

>Here we are; a couple of these during import:

I'm hoping/guessing these were already in the original database, and not added by this tool. That would be a bug on a whole different level. (But also unlikely, if the encryption code is not working I would expect a _lot_ more problems)

>And then Signal dies completely with:

Hm, something is definitely going wrong here. The tool seems to have written a 0 byte frame, not sure how that could happen though. I'm wondering if this backup could be read back by this program.

>OK, I fiddled with --autolimitdates so that it handled the desktop data being newer than the mobile backup (it only handled the other way), and that seems to have done it? No crash on import, and no huge database. There are 3–4 messages and missed calls that are phantom unread that were not before, and all of my conversations are starting on June, but it seems like things are generally back on track.

I adjusted `--autolimitdates` to now import everything that's newer than the newest message in the Android backup, as well as everything that is older than the oldest message. For cases where there are gaps in the message history in the middle, the `--limittodates` (non-auto) is still the way to go.

What do you mean exactly with 'all of my conversations are starting on June', is that a good thing?

---

If it's not too much trouble, maybe you could test your original command again (without the `--autolimitdates`) and see if the errors about the jsonlong-type are indeed solved. Also, you could then maybe open the resulting backup with this tool and see if it decodes it correctly (no need for any arguments, just assert that it reaches 100% on opening the backup). Lastly, maybe you could retry _with_ `--autolimitdates` to see it now correctly imports newer messages.

**Comment by sesse** [comment]
>> I got about 30 instances in my database, all in the same conversation. (I don't know why I didn't need to divide by 1000. Perhaps somehow my first calculation was wrong.)
> Timestamps in Signal (Android and desktop) are stored as milliseconds since epoch, not seconds.

Yes, but I originally thought the value worked out to microseconds. However, my initial calculation there was somehow fat-fingered.

>> I also got one of these, which was nonfatal:
> That's a weird one, it seems the author of the emoji-reaction is not found. Would it be possible for you to open your Signal Desktop instance and find this message? The reaction was posted October 16 2020, 18:58:24 CEST (but the message could be older). If this is too difficult (absolutely possible, if you have many active conversations around that period), I could write some instructions on how to more easily find it by running some queries on the database.

There's no good way I can get all conversations just listed by time, right? My guess is that I'd have to scroll through 20+ conversations, and Signal Desktop doesn't have a good way to fast-scroll… :-)

> Generally in these cases the attachment is missing (may have been deleted), I added a little more output in this case. Your database seems to be a bit different though, maybe the attachments' filename is simply under a different json-key? If you want to investigate further, let me know and I'll have the program output some more info at this point.

No, it's fine; I can live without 2–3 images.

>> So I assume something went wrong in the deduplication
> Ehm... what deduplication? :)

Haha, OK, I assumed there was some. I guess that explains why the database got so huge.

>> Here we are; a couple of these during import:
> I'm hoping/guessing these were already in the original database, and not added by this tool. That would be a bug on a whole different level. (But also unlikely, if the encryption code is not working I would expect a _lot_ more problems)

I don't know how the MAC system works at all, but none of them showed up when I added --autolimitdates, so they somehow have to be imported from the desktop database.

>> And then Signal dies completely with:
> Hm, something is definitely going wrong here. The tool seems to have written a 0 byte frame, not sure how that could happen though. I'm wondering if this backup could be read back by this program.

As I read the Java code (very briefly!): I think maybe it reads the MAC, sees that it fails verification, and then sends a 0-byte frame on to the rest of the code, which causes it to throw an exception (because it expects it to end in a 10-byte MAC). So in a sense, it's a bug in the Signal Android importer code, but when parsing what is somehow an invalid record.

> I adjusted `--autolimitdates` to now import everything that's newer than the newest message in the Android backup, as well as everything that is older than the oldest message. For cases where there are gaps in the message history in the middle, the `--limittodates` (non-auto) is still the way to go.

Sounds good!

> What do you mean exactly with 'all of my conversations are starting on June', is that a good thing?

When I go into a conversation on Android Signal, it's scrolled to June, as if that was the last time I read messages in that conversation. Which is wrong/weird, but only really cosmetic because I can just scroll down and it will be fine the next time. I guess Signal somehow stores a cursor?

> If it's not too much trouble, maybe you could test your original command again (without the `--autolimitdates`) and see if the errors about the jsonlong-type are indeed solved. Also, you could then maybe open the resulting backup with this tool and see if it decodes it correctly (no need for any arguments, just assert that it reaches 100% on opening the backup). Lastly, maybe you could retry _with_ `--autolimitdates` to see it now correctly imports newer messages.

Let me check.

**Comment by bepaald** [comment]
>There's no good way I can get all conversations just listed by time, right? My guess is that I'd have to scroll through 20+ conversations, and Signal Desktop doesn't have a good way to fast-scroll… :-)

Not that I know of. Would you be willing to run some sql queries for me? The program has a function to dump the unencrypted Signal Desktop database. It's not very user-friendly though: it still requires an Android backup as input (though it doesn't do anything with it after reading) and needs the Signal config directory twice, so:
```
signalbackup-tools [androidbackup] [passphrase] --dumpdesktopdb ~/.config/Signal/ ~/.config/Signal/
```

Those directories are the default location on Linux, on Windows it will be `$HOME/AppData/Roaming/Signal`.

After that you should find a sqlite database _"desktop.db"_ in your working directory. I think to get some more info on the message that the reaction belongs to you can run:
```
SELECT messages.* from messages, json_each(messages.json, '$.reactions') AS each WHERE json_extract(messages.json, '$.reactions') IS NOT NULL AND json_extract(each.value, '$.timestamp') IS 1602867504000;
```
(I often run `.mode line` first, to get the results in a slightly more readable format).

In the result should be a `body`, which is the message body (maybe this is already enough to find the message using Signal Desktop's search function). Also, there should be a `conversationId`. Running the following with this `conversationId` should show some info on the conversation (like the name of the contact/group):
```
SELECT * from conversations where id IS '05ecbab4-500a-484b-b207-ea903d1a2c79';
```

>I don't know how the MAC system works at all, but none of them showed up when I added --autolimitdates, so they somehow have to be imported from the desktop database.

Right, I actually find that worrying.

>As I read the Java code (very briefly!): I think maybe it reads the MAC, sees that it fails verification, and then sends a 0-byte frame on to the rest of the code, which causes it to throw an exception (because it expects it to end in a 10-byte MAC). So in a sense, it's a bug in the Signal Android importer code, but when parsing what is somehow an invalid record.

I don't think this is correct. The problem is that it tries to copy the last 10 bytes (= the MAC) to a small array to later check if it matches the calculated value. It does this by copying 10 bytes from the frame from position (frame.length - 10). However, it turns out that this frame.length = 0 and the position = -10 (as seen in the error message: `src.length=0 srcPos=-10`).

But I don't see how this could possibly happen: in my code, the length of a frame is set (here: https://github.com/bepaald/signalbackup-tools/blob/0d93af52c1db23fd0a49c9d8143aabed919dfd34/fileencryptor/encryptframe.cc#L60C1-L60C77) to an unsigned number (-> never below zero) + MACSIZE (== 10). So even if the frame size is 0, which would be a bug, the framelength would still be set to 10.

I'm honestly hoping that something simply went wrong with copying the large backup to the phone somehow. That could also explain those bad MACs on the attachment data. But maybe I'm just in denial about some huge bug at this point. 😅

>When I go into a conversation on Android Signal, it's scrolled to June, as if that was the last time I read messages in that conversation. Which is wrong/weird, but only really cosmetic because I can just scroll down and it will be fine the next time. I guess Signal somehow stores a cursor?

Ah, ok, that's news to me, I'll look into it at some point, but I also do not consider this a priority.

---

>The JSON dates still kill the import

So sorry about that, I forgot to copy in a few lines of code. Should be fixed now.

>As for --autolimitdates, it chose [...]

Looks good!

Thanks for reporting these problems and getting back with so quickly with so much useful info, that's really great!

**Comment by sesse** [comment]
> After that you should find a sqlite database _"desktop.db"_ in your working directory. I think to get some more info on the message that the reaction belongs to you can run:
> 
> ```
> SELECT messages.* from messages, json_each(messages.json, '$.reactions') AS each WHERE json_extract(messages.json, '$.reactions') IS NOT NULL AND json_extract(each.value, '$.timestamp') IS 1602867504000;
> ```

I found the message. It is a picture with a message, and there is indeed a heart emoji on it. Nothing looks bad at all in Signal Desktop.

> ```
> SELECT * from conversations where id IS '05ecbab4-500a-484b-b207-ea903d1a2c79';
> ```

That query came back completely empty. The conversation ID in the SQL for that message is 13338cbe-8445-418c-b073-a8a0c24bf732 (which also matches the fromId printed in the log).

>> I don't know how the MAC system works at all, but none of them showed up when I added --autolimitdates, so they somehow have to be imported from the desktop database.
> Right, I actually find that worrying.

I am inclined to agree :-)

> I don't think this is correct. The problem is that it tries to copy the last 10 bytes (= the MAC) to a small array to later check if it matches the calculated value. It does this by copying 10 bytes from the frame from position (frame.length - 10). However, it turns out that this frame.length = 0 and the position = -10 (as seen in the error message: `src.length=0 srcPos=-10`).

I don't know enough about the structure to help here, sorry.

> I'm honestly hoping that something simply went wrong with copying the large backup to the phone somehow. That could also explain those bad MACs on the attachment data. But maybe I'm just in denial about some huge bug at this point. 😅

FWIW, it happened twice on two different restores, so I don't think this is very likely.

> Ah, ok, that's news to me, I'll look into it at some point, but I also do not consider this a priority.

Not important, indeed.

>> The JSON dates still kill the import
> So sorry about that, I forgot to copy in a few lines of code. Should be fixed now.

Yes, works fine now.

**Comment by bepaald** [comment]
>I found the message. It is a picture with a message, and there is indeed a heart emoji on it. Nothing looks bad at all in Signal Desktop.

Ok, so I assume if you hover over it (or click it?) it shows the details with the author and all?

>That query came back completely empty. The conversation ID in the SQL for that message is 13338cbe-8445-418c-b073-a8a0c24bf732 (which also matches the fromId printed in the log).

Yes, sorry, of course it needed the actual conversationId (`13338cbe-8445-418c-b073-a8a0c24bf732`  in your case) instead of what I posted. So, if you look at the results of that query, do you see a `serviceId` and/or `e164`? Non-empty? Any other fields with a value in the same format (8-4-4-4-12)?

>FWIW, it happened twice on two different restores, so I don't think this is very likely.

Agreed. So, I think I have a bug somewhere. It's always very hard to debug these things remotely, so this will only go on as long as you are willing to keep answering my questions. I understand your problem is fixed (the import was more or less successful), so I completely understand if you're done. But, I am still very curious what happens if you let this program read the (large, 30Gb) backup.

Thanks!

**Comment by bepaald** [comment]
>It does.
The e164 shows the person's phone number; no serviceId. The only 8-4-4-4-12 field is the ID itself (in both the key, and in a JSON “id” field).

Thanks! Knowing this and looking closer at the code and your error message, this turns out to have been just a superfluous warning. It was already falling back to the e164, but still warning because of the empty uuid. I've changed it to only show that message when both uuid and e164 are empty.

>If you want to, I can probably give you supervised access; i.e., SSH access to a screen where you have access to both databases, but in a shared screen session so that I can verify that you're not grabbing all my messages or poking around to grab my friends' phone numbers. :-) It would be a bit cumbersome for both, though (and I can't give you access to a phone to import it into), so perhaps it's easier to do debugging-by-GitHub.
I gave the 30GB backup to the program, and it was entirely happy:

Yeah, that's a kind offer, but indeed a bit cumbersome. Also, I'd be almost as uncomfortable poking in someone else's database myself (even though I'm guessing I won't be able to read most of it).

I've added some more output (in verbose mode) to the program, maybe if you have time you could send me the logs, both of importing from desktop, as well as reading the merged file back.

(Assuming you're on Linux)
```
$ #log of import
$ ./signalbackup-tools [backupfile] [passphrase] --importfromdesktop --verbose -o mergedbackup | gzip > mergelog
$ #log of reading back
$ ./signalbackup-tools mergedbackup [passphrase] --verbose --fulldecode | gzip > readlog
```

The `gzip` is of course not technically necessary, but I expect both logs to be ~150Mb, so use any compression you prefer. I don't think there is any personal info in the output, but you could always email me directly instead of posting here if you want.

With these logs I should be able to definitely tell if there is a 0 byte frame in there somewhere, then I'll take it from there.

Thanks!

**Comment by sesse** [comment]
I have the files (and none of them seem to indicate a zero-frame, unfortunately); how do I get them to you privately?

**Comment by bepaald** [comment]
You can email them to me: basjetimmer [at] yahoo

Thanks!

**Comment by bepaald** [comment]
Ok, thanks.

I've checked the log (or, rather I've written a small program to do it for me), and as far as I can tell everything looks fine.

Obviously there was no plainly visible message saying '_Framelength: 0_', that would be easy to spot, but the fear was: what if the program just silently skips a 0 byte-sized frame?

However with these logs, calculating the reported framelengths and file positions (and attachment length where applicable), I must conclude the backup contains no frame with a size of 0. Such a frame would either make the `COUNTER` jump an unexpected number or mess up its value (which would render further decrypting impossible), or the numbers (position and length) would not add up correctly, resulting in (at least) a 4 byte difference somewhere (for the 4 byte `0` marker). None of this seems to happen.

Also, while the higher level functions (like `importFromDesktop()`) are a horrible, and complicated mess, the core of the program, the low level stuff (reading and decrypting the file and building the database) is all pretty well-written and easily readable (if I do say so myself), and I'm quite sure a 0 byte framesize would throw an error.

So, I'm afraid at this point, even though it is somewhat unlikely because it happened twice, I'm back to thinking the problem was some sort of corruption of the backup file. (I just remembered, I have heard of a similar situation when the storage is full on the phone: no proper error about full storage was shown, but NULL's where fed to Signal. Any chance that happened? The backup was certainly large enough :) )

Given the situation, the next step for me would be to repeat the backup-restore process very carefully by:

- first verifying the backup is ok on pc (with this tool)
- taking a hash of it (like MD5)
-  transferring to phone
- taking a hash on the phone to verify the backup is identical
- checking the storage space (maybe Signal needs _another_ 30Gb to build its internal database while restoring, I'm not sure)
- finally, perform the restore again (with `adb logcat` running).

However, since your problem is solved I would not ask you to go through all this trouble, unless you yourself really want to get to the bottom of this. It's up to you.

Thanks!
